common/app/conf/PanicSheddingFilter.scala

```diff
@@ -0,0 +1,157 @@
+package conf
+
+import java.lang.management.ManagementFactory
+
+import akka.agent.Agent
+import common.{ExecutionContexts, RequestMetrics, Logging}
+import conf.switches.Switches
+import org.joda.time.DateTime
+import play.api.mvc.Results._
+import play.api.mvc.{Result, RequestHeader, Filter}
+
+import scala.concurrent.Future
+
+// this turns requests away with 5xx errors if we are too busy
+object PanicSheddingFilter extends Filter with Logging {
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+  import scala.concurrent.duration._
+
+  val ALL_200s_MAX_LATENCY = 1.second.toMillis
+  val PANICING_MIN_LATENCY = 3.seconds.toMillis
+  // always allow a few concurrent connections even in slow times
+  val MIN_CONNECTIONS = 4
+  val MAX_STARTS_IMBALANCE = StartCompleteRatioMonitor.RANGE / 4
+
+  val averageLatency = Agent(LatencyMonitor.initialLatency)
+  val requestsInProgressCounter = Agent(InProgressRequestMonitor.initialRequestsInProgress)
+  val startCompleteRatio = Agent(StartCompleteRatioMonitor.initialRatio)
+
+  // Busy monitoring is not perfect as it only knows when things are already slow.
+  // if we get loads of requests at the same moment, it would take up to 1 second to realise
+  // and start throttling
+  // also if the request rate keeps going up there would be plenty of "young" requests to skew the average down
+  def available(request: RequestHeader) = {
+
+    import implicits.Requests._
+    // we read all the values up front, so there could be other requests ahead of us that would affect the values (race)
+    val previousLatency = averageLatency.get.latency
+    val requestsInProgress = requestsInProgressCounter.get
+    val startsRatio = startCompleteRatio.get.ratio
+    lazy val loadAverage = ManagementFactory.getOperatingSystemMXBean.getSystemLoadAverage
+    if (Switches.PanicLoggingSwitch.isSwitchedOn) {
+      log.info(s"$requestsInProgress in progress: averageLatency = $previousLatency, startsRatio (range is -100,100) = $startsRatio, loadAverage = $loadAverage")
+    }
+    if (startsRatio > MAX_STARTS_IMBALANCE) {
+      log.warn(s"Request spike detected, won't serve this request - $startsRatio (range is -100 to 100, over 25 is a spike), loadAverage = $loadAverage")
+      RequestMetrics.PanicRequestsSurgeMetric.increment()
+      false
+    } else if (previousLatency <= ALL_200s_MAX_LATENCY) {
+      true
+    } else if (previousLatency > PANICING_MIN_LATENCY) {
+      RequestMetrics.PanicExcessiveLatencyMetric.increment()
+      if (requestsInProgress <= MIN_CONNECTIONS) {
+        log.warn(s"Excessive previous latency detected, serving anyway as $requestsInProgress/$MIN_CONNECTIONS requests in progress - ${previousLatency}ms, loadAverage = $loadAverage")
```

Comment:
Should this one be more like `Excessive previous latency detected, serving anyway as request in progress ($requestsInProgress) is less than mininum connection ($MIN_CONNECTIONS) -  latency = ${previousLatency}ms, loadAverage = $loadAverage"`